src/inter_dc_dep_vnode.erl

```diff
@@ -0,0 +1,230 @@
+
+%% Update the clock value associated with the given DCID from the perspective of this partition.
+-spec update_clock(#state{}, dcid(), non_neg_integer()) -> #state{}.
+update_clock(State = #state{last_updated = LastUpdated, partition = Partition}, DCID, Timestamp) ->
+  %% Should we decrement the timestamp value by 1?
+  NewClock = vectorclock:set_clock_of_dc(DCID, Timestamp, State#state.vectorclock),
+
+  %% Check if the stable snapshot should be refreshed.
+  %% It's an optimization that reduces communication overhead during intensive updates at remote DCs.
+  %% This assumes that heartbeats/updates arrive on a regular basis,
+  %% and that there is always the next one arriving shortly.
+  %% This causes the stable_snapshot to tick more slowly, which is an expected behaviour.
+  Now = now_milisec(),
+  NewLastUpdated = case Now > LastUpdated + ?VECTORCLOCK_UPDATE_PERIOD of
+    %% Stable snapshot was not updated for the defined period of time.
+    %% Push the changes and update the last_updated parameter to the current timestamp.
+    %% WARNING: this update must push the whole contents of the partition vectorclock,
+    %% not just the current DCID/Timestamp pair in the arguments.
+    %% Failure to do so may lead to a deadlock during the connection phase.
+    true ->
+      %% Update the vectorclock the OLD way
+      vectorclock_vnode:update_partition_clock(Partition, NewClock),
+      vectorclock_vnode:recalculate_stable_snapshot(Partition),
+
+      %% Update the stable snapshot NEW way (as in Tyler's weak_meta_data branch)
+      %%ok = meta_data_sender:put_meta_dict(State#state.partition, NewClock),
+
+      Now;
+    %% Stable snapshot was recently updated, no need to do so.
+    false -> LastUpdated
+  end,
+
+  State#state{vectorclock = NewClock, last_updated = NewLastUpdated}.
+
+%% Get the current vectorclock from the perspective of this partition, with the updated entry for current DC.
+-spec get_partition_clock(#state{}) -> vectorclock().
+get_partition_clock(State) ->
+  %% Return the vectorclock associated with the current state, but update the local entry with the current timestamp
+  vectorclock:set_clock_of_dc(dc_utilities:get_my_dc_id(), inter_dc_utils:now_millisec(), State#state.vectorclock).
+
+%% Utility function: converts the transaction to a list of clocksi_payload ops.
+-spec updates_to_clocksi_payloads(#interdc_txn{}) -> list(#clocksi_payload{}).
+updates_to_clocksi_payloads(Txn = #interdc_txn{dcid = DCID, timestamp = CommitTime, snapshot = SnapshotTime}) ->
+  lists:map(fun(#operation{payload = LogRecord}) ->
+    {Key, Type, Op} = LogRecord#log_record.op_payload,
+    #clocksi_payload{
+      key = Key,
+      type = Type,
+      op_param = Op,
+      snapshot_time = SnapshotTime,
+      commit_time = {DCID, CommitTime},
+      txid =  LogRecord#log_record.tx_id
+    }
+  end, inter_dc_txn:ops_by_type(Txn, update)).
+
+now_milisec() ->
+  {MegaSecs, Secs, MicroSecs} = os:timestamp(),
```

Comment:
Should use dc_utiliities:now() for compatibility with other parts of the code (see https://github.com/SyncFree/antidote/pull/173 )